ci/checks/style.sh

```diff
@@ -82,4 +88,40 @@ else
   echo -e "\n\n>>>> PASSED: clang format check\n\n"
 fi
 
+# clang-tidy check
+# NOTE:
+#   explicitly pass GPU_ARCHS flag to avoid having to evaluate gpu archs
+# because there's no GPU on the CI machine where this script runs!
+# NOTE:
+#   also, sync all dependencies as they'll be needed by clang-tidy to find
+# relevant headers
+function setup_and_run_clang_tidy() {
+    local LD_LIBRARY_PATH_CACHED=$LD_LIBRARY_PATH
+    export LD_LIBRARY_PATH=$CONDA_PREFIX/lib:$LD_LIBRARY_PATH
+    mkdir cpp/build && \
+        cd cpp/build && \
+        cmake -DGPU_ARCHS=70 \
+              -DBLAS_LIBRARIES=${CONDA_PREFIX}/lib/libopenblas.so.0 \
+              .. && \
+        make treelite && \
```

Comment:
because clang-tidy requires all the headers to be available during its "compilation" time. And the only way we currently can get all these headers to be reliably made available to clang-tidy is to build all the cuml's dependencies.

Reply:
There has to be a better way to do this check. Between installing conda dependencies and make treelite, there is too much happening for this check.

Reply:
Unfortunately, no. This was the main reason I initially had proposed to make this as part of `ci/gpu/build.sh`. Because that CI already has all these dependencies handled as well as build files available in its cmake "build" folder.

Reply:
Clearly, there are reasons to not put it inside `ci/gpu/build.sh`. Instead, putting it inside `ci/check/style.sh` has shown it to be not "simple enough". Thus, how about we instead create a separate `ci/check/tidy.sh` CI just for this case? What say @raydouglass ?

Reply:
@dantegd @JohnZed @raydouglass can I get a resolution here please? This is the only item preventing us from adding this feature into cuML. Here's the summary:
1. For running clang-tidy, we need the compilation commands for each compilation unit in the project.
2. This is generated via `CMAKE_EXPORT_COMPILE_COMMANDS` option (which I have enabled by default in our cmake. It doesn't cost anything at all, except for one extra file in the "build" folder)
3. clang-tidy actually compiles the code. So, it requires all the headers to be available. This means, at the least, we'll have to build all of the dependencies inside `cpp/cmake/Dependencies.cmake`!

Even other projects like xgboost also end up [generating this compilation database](https://github.com/dmlc/xgboost/blob/master/tests/ci_build/tidy.py#L73) explicitly while running tidy check. However, unlike cuML, xgboost doesn't use `ExternalProject_Add`, so all it requires is to run cmake.

Reply:
And in case you want to know the runtime impact of adding this check into `style.sh`:
1. before: 26s (CI at commit 131e0cd above)
2. after: ~10m (CI at commit fd0fca5 above)

Reply:
BTW, here's my reasoning behind the `ci/check/tidy.sh` proposal:
1. It is a code style checker, but unlike other checkers, it requires atleast a partial build.
2. It requires build, but unlike `build.sh`, this doesn't have to be run on all OS, cuda and python versions.

Reply:
This check should stay where it is in `ci/check/style.sh` for now. Before this PR is merged, I'm going to make a small change to gpuCI so that a successful style check (and changelog) is required before running the other builds.

In the longer term, with the upcoming build process overhaul, we plan to build in a more robust check job infrastructure.

Reply:
cool. Thanks @raydouglass . Please do point us to any place where we can learn more about this new flow.

Reply:
also, I'm guessing that you are ok with this change to `style.sh`, atleast for now?